homeassistant/components/binary_sensor/zha.py

```diff
@@ -26,23 +30,48 @@
 
 async def async_setup_platform(hass, config, async_add_entities,
                                discovery_info=None):
-    """Set up the Zigbee Home Automation binary sensors."""
-    discovery_info = zha.get_discovery_info(hass, discovery_info)
+    """Old way of setting up Zigbee Home Automation binary sensors."""
+    pass
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Zigbee Home Automation binary sensor from config entry."""
+    async def async_discover(discovery_info):
+        await _async_setup_entity(hass, config_entry, async_add_entities,
+                                  discovery_info)
+
+    unsub = async_dispatcher_connect(
+        hass, ZHA_DISCOVERY_NEW.format(DOMAIN), async_discover)
+    hass.data[DATA_ZHA][DATA_ZHA_DISPATCHERS].append(unsub)
+
+    discovery_info = hass.data.get(DATA_ZHA, {})
+    binary_sensors = discovery_info.get('binary_sensor', None)
+    if binary_sensors is None:
+        return
+    for discovery_key in binary_sensors.keys():
+        await _async_setup_entity(hass, config_entry, async_add_entities,
+                                  discovery_key)
+
+
+async def _async_setup_entity(hass, config_entry, async_add_entities,
+                              discovery_info=None):
+    """Set up the ZHA binary sensor."""
+    discovery_info = zha.get_discovery_info(hass, 'binary_sensor',
```

Comment:
It doesn't look as if we should or need to call this. We already have all the info.